pkg/storegateway/bucket.go

```diff
@@ -1963,7 +1964,7 @@ func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
 }
 
 // NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
-func toPostingGroup(lvalsFn func(name string) ([]string, error), m *labels.Matcher) (*postingGroup, error) {
+func toPostingGroup(lvalsFn func(name string, filter func(string) bool) ([]string, error), m *labels.Matcher) (*postingGroup, error) {
```

Comment:
nit: I'd suggest to extract `func(name string, filter func(string) bool) ([]string, error)` into a separate type, to make it easier to see what is the type of `lvalsFn`, as this entire function signature is getting difficult to read.

Reply:
I've extracted a `labelValuesReader` interface that has only this function and now pass the indexHeader as an instance of that here: https://github.com/grafana/mimir/pull/3391/commits/e13c20f4a85f6a416bb9ba41e5890fd0e5c8c740